Add zlib-ng-compat port

netcorefan1 opened the pull request:
This is basically the same of the already existing `zlib-ng` port with the following additions:
`-DCMAKE_POSITION_INDEPENDENT_CODE=ON` (seems to be required to fix -fPIC linkage [issue](https://github.com/opencv/opencv/issues/22573#issuecomment-1263508538))
`-DZLIB_COMPAT=ON` (Compile with `zlib` compatible API)

As of now is possible to replace `zlib` dependency with `zlib-ng` through:
- An empty overlay port for `zlib` which depends on `zlib-ng`
- A custom overlay triplet that set `CMAKE_POSITION_INDEPENDENT_CODE` and `ZLIB_COMPAT` flags.

It works fine and packages that depends on the old `zlib` can now be compiled with the new `zlib-ng` and take advantage of next gen instructions with no changes to the source code.
However, once `zlib-ng` is compiled in this way, packages can't use `zlib-ng` natively anymore [#30870].

With this variant of the main `zlib-ng` port, the user can replace a `zlib` dependency with just an empty overlay that depends on `zlib-ng-compat` (no custom overlay triplet needed anymore) and the author of a port can still use `zlib-ng` natively. An application can install packages that depend on `zlib-ng` and packages that depend on `zlib-ng-compat` with no file name conflicts since the generated outputs are different.

**Please note**:
- I am not sure if vcpkg allows the user to handle this type of scenario without recurring to a port variant. So, before merging this pull request it may be better to see if there is a better alternative.

netcorefan1 committed:
Add zlib-ng-compat port

Allows to replace old zlib dependency with the new zlib-ng in compat mode without any change to the source code.

Neumann-A commented:
This can either be solved by adding a new port or providing a feature to `zlib-ng` which just builds the source twice with different configure options.

netcorefan1 commented on ports/zlib-ng-compat/portfile.cmake:
```diff
@@ -0,0 +1,31 @@
+        "-DZLIB_FULL_VERSION=${ZLIB_FULL_VERSION}"
+        -DZLIB_ENABLE_TESTS=OFF
+        -DWITH_NEW_STRATEGIES=ON
+        -DCMAKE_POSITION_INDEPENDENT_CODE=ON
```
Thanks. Feel free to edit the file. Does the original port needs the same correction?

netcorefan1 replied:
> this should come from the used toolchain and not be hardcoded into the portfile.

This is the same of the original port. So, a pull request should be sent with the correct value (which I don't know).

netcorefan1 replied:
I marked as resolved since I removed the offending option with commit bf31b4e0aa5985248e2ab6925cf29edb9ebe1715

netcorefan1 commented:
> This can either be solved by adding a new port or providing a feature to `zlib-ng` which just builds the source twice with different configure options.

Can you point me to a working sample? Just like a new port, this will still requires the submission of changes to the repo, right?

netcorefan1 committed:
Added "compat" mode features

netcorefan1 committed:
Handle zlib[compat] feature

netcorefan1 committed:
Merge branch 'microsoft:master' into master

dg0yt commented:
I don't see a a point in adding another alternative port which cannot be tested in vcpkg CI. 
If you want to have an original zlib-ng *and* another zlib-ng configuration overrding zlib, you can use a local overlay port for zlib which is not empty but implemented like your desired zlib-ng-compat.

netcorefan1 commented:
> I don't see a a point in adding another alternative port which cannot be tested in vcpkg CI. If you want to have an original zlib-ng _and_ another zlib-ng configuration overrding zlib, you can use a local overlay port for zlib which is not empty but implemented like your desired zlib-ng-compat.

I have done what @Neumann-A suggested me: a new port (which is the original pull request) or a feature to zlib-ng which just builds the source twice with different configure options (which are my new two commits).

> If you want to have an original zlib-ng and another zlib-ng configuration overrding zlib, you can use a local overlay port for zlib which is not empty but implemented like your desired zlib-ng-compat.

I tried this in all the possible ways and the only way I have been able to make it work is as you [said](https://github.com/microsoft/vcpkg/discussions/30870). It sounds to me more like an hack. A custom port inside the port folder mean that:
- I'll have to restart from scratch each time I reinstall vcpkg (or some vcpkg rebuild mechanism clean and restore that folder)
- I'll to manually adjust my port at each update
- And probably other side effects that now don't come in mind

I think that a `compat` feature like suggested from Neumann should solve the problems above by keeping everything synchronized and without any additional port. People can also benefit from this by using high speed offered in compat mode with no change to the source and with just the effort of a local overlay.

netcorefan1 reviewed (commented):
If I have understood well, `-DCMAKE_POSITION_INDEPENDENT_CODE` should be removed and let the toolchain to take the decision. Let me see if there is a way to change the file...

netcorefan1 committed:
Removed -DCMAKE_POSITION_INDEPENDENT_CODE

As suggested here: https://github.com/microsoft/vcpkg/pull/30918#discussion_r1168587403

netcorefan1 committed:
Removed -DCMAKE_POSITION_INDEPENDENT_CODE 

As suggested here: https://github.com/microsoft/vcpkg/pull/30918#discussion_r1168587403

BillyONeal reviewed (changes requested):
This conflicts with the zlib port and as a result I don't believe it can be in the curated registry.

netcorefan1 commented:
@BillyONeal Unless you didn't found a way on how to solve the issues, I think I should close this pull request. I have published a [tutorial](https://github.com/microsoft/vcpkg/discussions/31116) (that needs approval and fixes from experts like you) for people that still want to implement this locally.
Thanks

BillyONeal commented:
Thanks for your contribution to vcpkg! We are sorry that we are unable to accurately model this alternative situation without overlay ports right now. Perhaps someday we will get better modeling of this situation and we'll be able to add something like this back.

Thanks again and have a great day!
